Fix --nv option

mrmhodak opened the pull request:
--nv option needs to be adjusted because it "imports" several CUDA libraries, which should be provided by the container.

This commit does two things:
1. Remove two unnecessary entries from the "grep" pattern:
   a) "libnvcu", because "libnv" already catches that
   b} "libOpenCL". Strangely, both kernel driver and cuda  provide libOpenCL.so (installed in two different locations), which differ according to diff. However, the cuda one has a higher priority according to ldconfig, and so OpenCL already in the container should be used. Because of this the grep pattern should not include libOpenCL.
2. Because grep catches several cuda-related libraries, specifically libcudart.so*, libnvrtc.so, and libnvToolsExt.so* I am adding grep -v statement after the pipe to avoid this.

One way this issue manifested itself, was that libraries in /.singularity.d/libs were replicated with libraries inside the container. With this fix, it does not happen anymore.

Finally, this is related to Pull Request #774 which I am going to delete.

Attn: @singularityware-admin

Commit:
Fix --nv option

--nv option needs to be adjusted because it "imports" several CUDA libraries, which should be provided by the container.

This commit does two things:
1. Remove two unnecessary entries from the "grep" pattern:
   a) "libnvcu", because "libnv" already catches that
   b} "libOpenCL". Strangely, both kernel driver and cuda  provide libOpenCL.so (installed in two different locations), which differ 
         according to diff. However, the one from cuda has a higher priority according to ldconfig, which means that OpenCL from the 
         container should be used. Because of this the grep pattern should not include libOpenCL.
2. Because grep catches several cuda-related libraries, specifically libcudart.so*, libnvrtc.so, and libnvToolsExt.so* I am adding grep -v statement after the pipe to avoid this.

GodloveD committed:
making nvidia libs configurable by user

Commit:
Merge pull request #1 from GodloveD/nv_fix

making nvidia libs configurable by user

gmkurtzer commented on libexec/cli/action_argparser.sh:
```diff
@@ -116,22 +116,24 @@ while true; do
-                    else
-                        SINGULARITY_CONTAINLIBS="$SINGULARITY_CONTAINLIBS,$i"
+            for lib in $(cat $SINGULARITY_sysconfdir"/singularity/nvliblist.conf" | grep -v "^#"); do
+                for i in $(ldconfig -p | grep "${lib}"); do
```
I would prefer this being done with a single grep like `grep -f "$SINGULARITY_sysconfdir"/singularity/nvliblist.conf"` instead of nested loops.

GodloveD replied:
I'm not sure if I follow what you're thinking.  The first loop iterates over the library patterns in nvliblist.conf while the second loop iterates over the matches in the ld.so.cache.  Sorry if I'm missing something easy, but I don't see how it could be done with a single loop.

gmkurtzer replied:
By using `ldconfig -p | grep -f...` and then looping over the output of that, you can omit the outer loop completely.

mrmhodak replied:
That would work, but the comments and empty lines need to be stripped first

GodloveD replied:
Ah.  Thanks for the clarification.  I see what you mean.  I just implemented it.  Will fire off another pull request to @mrmhodak in a minute.

Commit:
Replacing nested loop by grep -f 

Porting contribution by @GodloveD to the right branch

GodloveD commented on libexec/cli/action_argparser.sh:
```diff
@@ -116,22 +116,25 @@ while true; do
             shift
-            for i in `ldconfig -p | grep -E "/libnv|/libcuda|/libEGL|/libGL|/libnvcu|/libvdpau|/libOpenCL|/libOpenGL"`; do
+            SINGULARITY_NVLIBLIST=$(mktemp singularity-nvliblist.XXXXXXXX)
+            cat $SINGULARITY_sysconfdir"/singularity/nvliblist.conf" | grep -v "^#" > $SINGULARITY_NVLIBLIST
```
This leaves an empty line.  Which will apparently match everything (`\n') in the output of `ldconfig -p`. YUCK!

GodloveD replied:
I think it needs to be:
```bash
cat $SINGULARITY_sysconfdir"/singularity/nvliblist.conf" | grep -Ev "^#|^\s*$"
```
I'm going to test a whole bunch more though.

mrmhodak replied:
This looks like a right way to fix the issue and my testing confirms that it works. Adding into the branch.

Commit:
Fix grep in action_argparser.sh

Previous version left an empty line in the grep matching pattern, fixing it now.
